### PR TITLE
uniffi-bindgen: init at 0.17.0

### PR DESCRIPTION
**Pull request by vtuan10**
###### Motivation for this change
I want to use uniffi-bindgen to generate language bindings for kotlin, swift, python and ruby from my rust code.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/uniffi-bindgen/default.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+rustPlatform.buildRustPackage rec {
+  pname = "uniffi-bindgen";
+  version = "0.17.0";
+
+  src = fetchFromGitHub {
+    owner = "mozilla";
+    repo = "uniffi-rs";
+    rev = "01a9266cef151e57b2d1f98f0dbfd524fc2c72ea"; # no tag for this release yet
+    sha256 = "sha256-7/3J77xMdx8xhrlUSpBQyU5uOKH6q/AnYA8kqEMJxRQ=";
```
Comment (veehaitch) [inline pkgs/development/tools/uniffi-bindgen/default.nix:22]:
```suggestion
    hash = "sha256-7/3J77xMdx8xhrlUSpBQyU5uOKH6q/AnYA8kqEMJxRQ=";
```

File: pkgs/development/tools/uniffi-bindgen/default.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+, pkg-config
+, ktlint
+, yapf
+, rubocop
+, rustfmt
+, libiconv
+, makeWrapper
+}:
+
+rustPlatform.buildRustPackage rec {
```
Comment (veehaitch) [inline pkgs/development/tools/uniffi-bindgen/default.nix:14]:
```suggestion
rustPlatform.buildRustPackage {
```

**Review by veehaitch — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/uniffi-bindgen/default.nix (hunk 1 above)
Comment (veehaitch) [inline pkgs/development/tools/uniffi-bindgen/default.nix:22]: Upstream just pushed the tag: https://github.com/mozilla/uniffi-rs/pull/1163#issuecomment-1040353354

```suggestion
    rev = "v${version}";
    hash = "sha256-EGyJrW0U/dnKT7OWgd8LehCyvj6mxud3QWbBVyhoK4Y=";
```

**Review by veehaitch — CHANGES_REQUESTED** [review]

**Comment by veehaitch** [comment]
Result of `nixpkgs-review pr 159617` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>uniffi-bindgen</li>
  </ul>
</details>

**Review by veehaitch — APPROVED** [review]

File: pkgs/development/tools/uniffi-bindgen/default.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  version = "0.17.0";
+
+  src = fetchFromGitHub {
+    owner = "mozilla";
+    repo = "uniffi-rs";
+    rev = "v${version}";
+    hash = "sha256-EGyJrW0U/dnKT7OWgd8LehCyvj6mxud3QWbBVyhoK4Y=";
+  };
+
+  cargoPatches = [ ./cargo-lock.patch ];
```
Comment (SuperSandro2000) [inline pkgs/development/tools/uniffi-bindgen/default.nix:25]:
```suggestion
  cargoLock.lockFile = ./Cargo.lock;
```
Please use the lock file directly.

File: pkgs/development/tools/uniffi-bindgen/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  cargoHash = "sha256-Fw+yCAI32NdFKJGPuNU6t0FiEfohoVD3VQfInNJuooI=";
+
+  cargoBuildFlags = [ "-p uniffi_bindgen" ];
+  cargoTestFlags = [ "-p uniffi_bindgen" ];
+  nativeBuildInputs = [ pkg-config makeWrapper ];
```
[the remark this answers is not in the document]
Comment (vtuan10) [reply pkgs/development/tools/uniffi-bindgen/default.nix]: Thanks! I removed it and also removed `libiconv` as well, as that is already added as `buildInput` in `buildRustPackage`.

File: pkgs/development/tools/uniffi-bindgen/default.nix (hunk 3 above)
Comment (vtuan10) [reply pkgs/development/tools/uniffi-bindgen/default.nix]: Seems like `cargoLock.lockFile` does not work to generate a lockfile. I changed it to `cargoLock.lockFileContents` and added a `cp` in the `postPatch` as both options do not add a `Cargo.lock`.

**Pull request merged**